Enhance Concept Docs

nkakkar81 opened the pull request:
Enhanced concept docs for Pipeline, Pipelinerun, Taskrun, Task, Workspaces and Triggers.

vdemeester commented on modules/op-about-pipelines.adoc:
```diff
@@ -0,0 +1,65 @@
+
+A _Pipeline_ is a collection of Tasks, arranged in a specific order of execution, using which you can define the Continuous Integration/Continuous Delivery (CI/CD) work-flow for your applications.
+
+A Pipeline definition consists of a number of fields or attributes which together enable the Pipeline to accomplish a specific goal. Each Pipeline definition must contain a Task(s) which ingests specific inputs and produces specific outputs. The Pipeline definition can also optionally include Conditions, Workspaces, Parameters or Resources depending on the application requirements.
```
should "must contain a Task(s)" be "must contain at least a Task" ?

nkakkar81 commented:
@praveen4g0 @ppitonak - Request you to please review this PR and provide comments.

ppitonak commented on modules/op-about-taskrun.adoc:
```diff
@@ -0,0 +1,38 @@
+[id="about-taskrun_{context}"]
+= About TaskRun
+
+A _TaskRun_ instantiates a Task for execution with specific inputs, outputs, and execution parameters on a cluster. It can be invoked on its own or as part of a Pipeline.
```
@vdemeester Isn't `...as part of a PipelineRun` more precise?

vdemeester replied:
Yeah, as we are talking about `TaskRun`, I would use `PipelineRun` here.

ppitonak commented on modules/op-about-taskrun.adoc:
```diff
@@ -0,0 +1,38 @@
+
+A Task consists of one or more Steps that execute container images, and each container image performs a specific piece of build work. A TaskRun executes the Steps in a Task in the specified order, till all Steps execute successfully or a failure occurs.
+
+Here is an example of a TaskRun to run a Pipeline `build-and-deploy` with relevant input and output parameters:
```
This looks like a copy&paste... this run a Task, not a Pipeline called `apply-manifests` with relevant input (and param), there are no outputs

ppitonak commented on modules/op-about-pipelines.adoc:
```diff
@@ -0,0 +1,65 @@
+...
+----
+<1> Pipeline API version `v1beta1`.
+<2> Specifies the type of Kubernetes object, in this case `PipelineRun`.
```
PipelineRun -> Pipeline

ppitonak commented on modules/op-about-pipelinerun.adoc:
```diff
@@ -0,0 +1,38 @@
+
+All the Tasks in the Pipeline are executed in the defined sequence until all Tasks are successful or a Task fails. The `status` field tracks and stores the progress of each TaskRun in the PipelineRun for monitoring and auditing purpose.
+
+Here is an example of a PipelineRun to run a Pipeline `build-and-deploy` with relevant input and output parameters:
```
`input and outpu parameters` doesn't sound correct... I would suggest simply `resources and parameters`

nkakkar81 commented:
> Overall, this a good content. However, it feels a bit chaotic because we mix core concepts and how-to guide in one document. As a new user, I want to get familiar with terminology and basic concepts first, I expect to find it (directly or linked) in the very first document. Then, I want to try our what I learned.
> 
> The document "Understanding OpenShift Pipelines" already contains a short description of some kinds of resources. What about moving all of these "About X" section into that document and keeping "Creating applications with OpenShift Pipelines" as a kind of getting-started guide?

@Preeticp - I agree with Pavol is saying above, however I am not very sure if I should add this info to the "Understanding Pipelines" section as is , as the definitions ( in Concept topic) might seem redundant. I was thinking of revamping the Understanding Section, by removing the concepts topics ( one-liner definitions), and instead adding these more detailed modules to "Understanding Pipelines" section. Let me know what you think, We can also discuss this in our 1-1 in more detail.

nkakkar81 commented:
@vdemeester @ppitonak  - I have modified this PR to include concept information on Workspaces, and have also added example for Triggers. I have also restructured the Understanding Pipelines section to include all Concept topics. The only thing that I am still contemplating is whether I should keep the heading as "About XXX" or Just "XXX". For example, the heading of the topic for Tasks should be "About Tasks" or just "Tasks". Please let me know if you have any preference. 

Request you to please review this PR, and let me know your comments. Thanks

VeereshAradhya commented on modules/op-about-workspace.adoc:
```diff
@@ -0,0 +1,96 @@
+   - name: source <4>
+     workspace: shared-workspace <5>
+   runAfter:
+   - fetch-repository
```
there is no task with the name `fetch-repository`

vdemeester commented on modules/op-about-workspace.adoc:
```diff
@@ -0,0 +1,96 @@
+[id="about-workspaces_{context}"]
+= About Workspaces
+
+Workspaces declare shared storage volumes that a Task in a Pipeline needs at runtime. Instead of specifying the actual location of the volumes, Workspaces enable the Task authors to declare the filesystem or parts of the filesystem that would be required at runtime. The onus is on the Task or Pipeline user to provide the specific details of the volume that is mounted into that Workspace in a Task ...
```
The onus -> The bones :stuck_out_tongue_closed_eyes:

nkakkar81 replied:
replaced and reworded :)

ppitonak commented on modules/op-about-triggers.adoc:
```diff
@@ -19,3 +19,88 @@ Triggers consist of the following main components that work together to form a r
 
 //image::op-triggers.png[]
+
+Here is an code snippet of the `vote-app-binding` TriggerBinding which extracts Git repository information from the received event payload:
```
a code snippet

ppitonak commented on modules/op-about-triggers.adoc:
```diff
@@ -19,3 +19,88 @@ Triggers consist of the following main components that work together to form a r
+    metadata:
+      name: $(params.git-repo-name)-git-repo-$(uid) <7>
+    spec:
+      type: git <8>
```
`<7>` and `<8>` are not explained below

Preeticp commented:
Thanks Nidhi, the YAML examples are very well explained. I have added a few minor comments, PTAL.

nkakkar81 commented on modules/op-about-pipelines.adoc:
```diff
@@ -0,0 +1,65 @@
+[id="about-pipelines_{context}"]
+= Pipelines
+
+A _Pipeline_ is a collection of Tasks, arranged in a specific order of execution, using which you can define the Continuous Integration/Continuous Delivery (CI/CD) work-flow for your applications.
```
removed the expanded form

nkakkar81 commented:
@Preeticp  - I have incorporated your comments, request you to PTAL and provide /lgtm if you are ok with the changes.

Preeticp commented:
/lgtm

ahardin-rh commented on modules/op-about-pipelinerun.adoc:
```diff
@@ -0,0 +1,38 @@
+<1> PipelineRun API version `v1beta1`.
+<2> Specifies the type of Kubernetes object, in this case `PipelineRun`.
+<3> Unique name to identify this PipelineRun.
+<4> Name of the Pipeline to be run, in this case `build-and-deploy`.
```
run. In this example, `build-and-deploy`.

ahardin-rh commented on modules/op-about-pipelines.adoc:
```diff
@@ -0,0 +1,65 @@
+
+A _Pipeline_ is a collection of Tasks, arranged in a specific order of execution, using which you can define the CI/CD workflow for your applications.
+
+A Pipeline definition consists of a number of fields or attributes which together enable the Pipeline to accomplish a specific goal. Each Pipeline definition must contain at least one Task which ingests specific inputs and produces specific outputs. The Pipeline definition can also optionally include Conditions, Workspaces, Parameters, or Resources depending on the application requirements.
```
attributes, which together

ahardin-rh replied:
one Task, which

ahardin-rh commented on modules/op-about-pipelines.adoc:
```diff
@@ -0,0 +1,65 @@
+[id="about-pipelines_{context}"]
+= Pipelines
+
+A _Pipeline_ is a collection of Tasks, arranged in a specific order of execution, using which you can define the CI/CD workflow for your applications.
```
This first sentence is a little unclear and I don't think it will translate well. The incorporation of "using which" is confusing to me. Maybe start a new sentence: Use a pipeline to define the CI/CD workflow for your applications.

nkakkar81 replied:
Have modified with this sentence: A _Pipeline_ is a collection of Tasks arranged in a specific order of execution. You can define a CI/CD workflow for your application using Pipelines containing one or more Tasks.

ahardin-rh commented on modules/op-about-pipelines.adoc:
```diff
@@ -0,0 +1,65 @@
+...
+----
+<1> Pipeline API version `v1beta1`.
+<2> Specifies the type of Kubernetes object, in this case `Pipeline`.
```
object. In this example, `Pipeline`.

ahardin-rh commented on modules/op-about-pipelines.adoc:
```diff
@@ -0,0 +1,65 @@
+<4> Specifies the definition and structure of the Pipeline.
+<5> Set of resources required to run the Tasks specified in the Pipeline.
+<6> Parameters used across all the Tasks in the Pipeline.
+<7> Specifies the list of Tasks used in the Pipeline. Here two Tasks are specified - `build-image` and `apply-manifests`.
```
In this example, two Tasks are specified: `build-image` and `apply-manifests`.

ahardin-rh commented on modules/op-about-pipelines.adoc:
```diff
@@ -0,0 +1,65 @@
+<5> Set of resources required to run the Tasks specified in the Pipeline.
+<6> Parameters used across all the Tasks in the Pipeline.
+<7> Specifies the list of Tasks used in the Pipeline. Here two Tasks are specified - `build-image` and `apply-manifests`.
+<8> Task `build-image` which uses the `buildah` ClusterTask to build application images from a given Git repository.
```
build-image`, which
`apply-manifests`, which

ahardin-rh commented on modules/op-about-pipelines.adoc:
```diff
@@ -0,0 +1,65 @@
+<7> Specifies the list of Tasks used in the Pipeline. Here two Tasks are specified - `build-image` and `apply-manifests`.
+<8> Task `build-image` which uses the `buildah` ClusterTask to build application images from a given Git repository.
+<9> Task `apply-manifests` which uses a user-defined Task with the same name.
+<10> Specifies the sequence in which Tasks are run in a Pipeline. Here the `apply-manifests` Task is run only after the `build-image` Task is completed.
```
In this example,  the `apply-manifests`

ahardin-rh commented on modules/op-about-taskrun.adoc:
```diff
@@ -0,0 +1,38 @@
+A _TaskRun_ instantiates a Task for execution with specific inputs, outputs, and execution parameters on a cluster. It can be invoked on its own or as part of a PipelineRun.
+
+A Task consists of one or more Steps that execute container images, and each container image performs a specific piece of build work. A TaskRun executes the Steps in a Task in the specified order, till all Steps execute successfully or a failure occurs.
+
```
s/till/until

ahardin-rh commented on modules/op-about-taskrun.adoc:
```diff
@@ -0,0 +1,38 @@
+    kind: Task
+----
+<1> TaskRun API version `v1beta1`.
+<2> Specifies the type of Kubernetes object, in this case `TaskRun`.
```
object. In this example, `TaskRun`.

ahardin-rh commented on modules/op-about-taskrun.adoc:
```diff
@@ -0,0 +1,38 @@
+<1> TaskRun API version `v1beta1`.
+<2> Specifies the type of Kubernetes object, in this case `TaskRun`.
+<3> Unique name to identify this TaskRun.
+<4> Definition of the TaskRun. For this TaskRun, a set of input resources, parameters and Tasks are specified.
```
parameters, and

ahardin-rh commented on modules/op-about-tasks.adoc:
```diff
@@ -33,6 +33,11 @@ spec:
     - install
 ----
+<1> Task API version `v1beta1`.
+<2> Specifies the type of Kubernetes object, in this case `Task`.
```
object. In this example,  `Task`.

ahardin-rh commented on modules/op-about-triggers.adoc:
```diff
@@ -5,18 +5,103 @@
+----
+
+<1> TriggerBinding API version `v1alpha1`.
+<2> Specifies the type of Kubernetes object, in this case `TriggerBinding`.
```
object. In this example. `TriggerBinding`.

ahardin-rh commented on modules/op-about-triggers.adoc:
```diff
@@ -5,18 +5,103 @@
+<1> TriggerBinding API version `v1alpha1`.
+<2> Specifies the type of Kubernetes object, in this case `TriggerBinding`.
+<3> Unique name to identify this TriggerBinding.
+<4> List of parameters which will be extracted from the received event payload and passed to the TriggerTemplate. Here, the Git repository URL, name, and revision are extracted from the body of the event payload.
```
In this example, the Git

ahardin-rh commented on modules/op-about-triggers.adoc:
```diff
@@ -5,18 +5,103 @@
+----
+
+<1> TriggerTemplate API version `v1alpha1`.
+<2> Specifies the type of Kubernetes object, in this case `TriggerTemplate`.
```
object. In this example,

ahardin-rh commented on modules/op-about-triggers.adoc:
```diff
@@ -5,18 +5,103 @@
+      name: vote-app-template
+----
+<1> EventListener API version `v1alpha1`.
+<2> Specifies the type of Kubernetes object, in this case `EventListener`.
```
object. In this example,

ahardin-rh commented on modules/op-about-workspace.adoc:
```diff
@@ -0,0 +1,96 @@
+* A PersistentVolumeClaim from a provided VolumeClaimTemplate
+* An emptyDir that is discarded when the TaskRun completes
+
+Here is a code snippet of the `build-and-deploy` Pipeline which declares a `shared-workspace` Workspace for the Tasks, `build-image` and `apply-manifests`, defined in the Pipeline.
```
`build-image` , and

nkakkar81 replied:
I think we can avoid this comma, as there is a comma just before `build-image`.

ahardin-rh commented on modules/op-about-workspace.adoc:
```diff
@@ -0,0 +1,96 @@
+    - build-image
+...
+----
+<1> List of Workspaces shared between the Tasks defined in the Pipeline. A Pipeline can define as many Workspaces as required. Note that here only one Workspace named `shared-workspace` is declared.
```
s/Note that here/In this example,

ahardin-rh commented on modules/op-about-workspace.adoc:
```diff
@@ -0,0 +1,96 @@
+<5> Name of the Pipeline Workspace used by the Task. Note that the Workspace `source` in turn uses Pipeline Workspace `shared-workspace`.
+<6> List of Workspaces used in `apply-manifests` Task. Note that this Task shares the `source` Workspace with `build-image` Task.
+
+Here is a code snippet of the `build-deploy-api-pipelinerun` PipelineRun which uses a PersistentVolumeClaim for defining the storage volume for the `shared-workspace` Workspace used in the `build-and-deploy` Pipeline.
```
PipelineRun, which

ahardin-rh commented on modules/op-about-workspace.adoc:
```diff
@@ -0,0 +1,96 @@
+----
+<1> Specifies the list of Pipeline Workspaces for which volume binding will be provided in the PipelineRun.
+<2> The name of the Workspace in the Pipeline for which the volume is being provided.
+<3> Specifies the name of a predefined PersistentVolumeClaim which will be attached to the Workspace. Here an existing `source-pvc` PersistentVolumeClaim is attached with the `shared-workspace` Workspace.
```
, which will 
s/Here/In this example,

ahardin-rh commented on modules/op-pipelines-concepts.adoc:
```diff
@@ -17,6 +17,8 @@ TaskRun:: A TaskRun is automatically created by a PipelineRun for each Task in a
 
 PipelineResource:: A PipelineResource is an object that is used as an input and output for Pipeline Tasks. For example, if an input is a Git repository and an output is a container image built from that Git repository, these are both classified as PipelineResources. PipelineResources currently support Git resources, Image resources, Cluster resources, Storage Resources and CloudEvent resources.
 
+Workspace:: A Workspace is a storage volume that a Task requires at runtime to receive input or provide output. A Task or Pipeline declares the Workspace, and a TaskRun or PipelineRun provides the actual location of the storage volume which mounts on the declared Workspace. This makes the Task flexible, reusable and allows the Workspaces to be shared across multiple Tasks.
```
volume, which

ahardin-rh replied:
reusable, and

ahardin-rh commented:
The updates look great!

ahardin-rh commented:
/cherrypick enterprise-4.5

ahardin-rh commented:
/cherrypick enterprise-4.6
